src/downloader/worker.rs

```
@@ -0,0 +1,120 @@
+                    .unwrap_or_else(|| "Unknown Error".to_string()),
+            }));
+            return;
+        }
+
+        if attempt > 0 {
+            req.retry();
+            // Basic exponential backoff
+            let delay_ms = 1000 * 2u64.pow(attempt as u32 - 1);
+            let delay = Duration::from_millis(delay_ms);
+
+            tokio::select! {
+                _ = tokio::time::sleep(delay) => {},
+                _ = req.cancel.cancelled() => {
+                    req.cancel();
+                    return;
+                }
+            }
+        }
+
+        match download(client.clone(), &mut req).await {
+            Ok(file) => {
+                req.complete(file);
+                return;
+            }
+            Err(e) => {
+                if should_retry(&e) {
+                    last_error = Some(e);
+                    continue;
+                }
+
+                req.fail(e);
+                return;
+            }
+        }
+    }
+}
+
+async fn download(client: Client, req: &mut DownloadRequest) -> Result<File, Error> {
+    let mut response = client
+        .get(req.url().as_ref())
+        .send()
+        .await?
+        .error_for_status()?;
+    let total_bytes = response.content_length();
+    let mut bytes_downloaded = 0u64;
+
+    // Create the destination directory if it doesn't exist
+    if let Some(parent) = req.destination().parent() {
+        tokio::fs::create_dir_all(parent).await?;
+    }
+    let mut file = File::create(&req.destination()).await?;
+
+    let mut progress = DownloadProgress::new(bytes_downloaded, total_bytes);
+    req.send_progress(progress);
+
+    loop {
+        tokio::select! {
+            _ = req.cancel.cancelled() => {
+                drop(file); // Manually drop the file handle to ensure that deletion doesn't fail
+                tokio::fs::remove_file(&req.destination()).await?;
+                return Err(Error::Download(DownloadError::Cancelled));
```

Perhaps ignore any errors from `remove_file` so cancellation doesn’t hide the real problem

Same at L106‑108

---

We should log the errors instead of ignoring them. We need a logger.